plugins/exampleplugin/exampleplugin.cpp

```diff
@@ -0,0 +1,67 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2016 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "platform.h"
+#include "eclrtl.hpp"
+#include "jstring.hpp"
+#include "exampleplugin.hpp"
+
+#define EXAMPLE_PLUGIN_VERSION "example-plugin plugin 1.0.0"
+ECL_EXAMPLE_PLUGIN_API bool getECLPluginDefinition(ECLPluginDefinitionBlock *pb)
+{
+    /*  Warning:    This function may be called without the plugin being loaded fully.
+     *              It should not make any library calls or assume that dependent modules
+     *              have been loaded or that it has been initialised.
+     *
+     *              Specifically:  "The system does not call DllMain for process and thread
+     *              initialization and termination.  Also, the system does not load
+     *              additional executable modules that are referenced by the specified module."
+     */
+
+    if (pb->size != sizeof(ECLPluginDefinitionBlock))
+        return false;
+
+    pb->magicVersion = PLUGIN_VERSION;
+    pb->version = EXAMPLE_PLUGIN_VERSION;
+    pb->moduleName = "lib_redis";
+    pb->ECL = NULL;
+    pb->flags = PLUGIN_IMPLICIT_MODULE;
```

Comment:
Nope, sorry, but doing a file search for them suggests:

PLUGIN_IMPLICIT_MODULE & PLUGIN_DLL_MODULE are for backward compatibility - https://github.com/hpcc-systems/HPCC-Platform/blob/master/ecl/hql/hqlcollect.cpp#L599

PLUGIN_MULTIPLE_VERSIONS doesn't seem to be referenced anywhere and looks unused other than to fall into the last else block in the IF of the above link.

ZOMBIE_MODULE is only used here https://github.com/hpcc-systems/HPCC-Platform/blob/master/ecl/hql/hqlcollect.cpp#L1051  but am not sure what that implies. It is also not sued in any of the OSS plugins, perhaps these are for some of the LN ones? Yep, it is present more in LN and looks archive repo and legacy related.

PLUGIN_SAVEMASK seems to have something to do with saving the flags for the definition.

I'm only taking loose guesses here, it may be better to ask this ion flowdock.